### ISSUE TITLE
Update lastLogMark when replaying journal

### ISSUE THREAD

**Issue by karanmehta93** [apache/bookkeeper#2087]
**QUESTION**

Consider the following scenario.

Bookie 1 started cleanly.
Bookie replays journals (there aren't any) and flushed `lastLogMark` to disk `(logFileId, logFileOffset) -> (0, 0)`
5 entries are written to it (Journal fsync done but entry logs are not flushed, hence `lastLogMark` is not updated on disk) `(logFileId, logFileOffset) -> (15156416854, 4096)`
Bookie crashes.
Bookie restarts in read only mode.
Bookie reads the persisted `lastLogMark` as `(logFileId, logFileOffset) -> (0, 0)`
Bookie replays journals but doesn't update `lastLogMark` while doing it (since in current behavior it gets updated only on journal writes)
Bookie flushes the ledger and flushes the `lastLogMark`
However the value written again is `(logFileId, logFileOffset) -> (0, 0)` and not the updated one.

Is this behavior expected? 
In my understanding, the `lastLogMark` field should get updated even when replaying the journals so that the new value can be persisted to disk.

@jvrao @reddycharan @sijie Thoughts?

This is not a huge issue if the bookie restarts in RW mode, since the `lastLogMark` will get updated as soon as bookie receives its first write and `syncThread` work will go on.

**Comment by karanmehta93** [comment]
I wrote up this test for `BookieInitializationTest` class to corroborate my understanding of the code. There are 3 different lines that print the checkpoint to console.

```
    @Test
    public void someTest() throws Exception {
        File tmpDir = createTempDir("DiskCheck", "test");

        final ServerConfiguration conf = newServerConfiguration()
                .setJournalDirName(tmpDir.getPath())
                .setLedgerDirNames(new String[] { tmpDir.getPath() })
                .setDiskCheckInterval(1000)
                .setLedgerStorageClass(SortedLedgerStorage.class.getName())
                .setAutoRecoveryDaemonEnabled(false)
                .setZkTimeout(5000);

        BookieServer server = new MockBookieServer(conf);
        server.start();
        Checkpoint checkpoint = server.getBookie().journals.get(0).newCheckpoint();
        System.out.println("Checkpoint: " + checkpoint);
        ClientConfiguration clientConf = new ClientConfiguration();
        clientConf.setMetadataServiceUri(metadataServiceUri);
        BookKeeper bkClient = new BookKeeper(clientConf);
        LedgerHandle lh = bkClient.createLedger(1, 1, 1, DigestType.CRC32, "passwd".getBytes());
        long entryId = -1;
        long numOfEntries = 5;
        for (int i = 0; i < numOfEntries; i++) {
            entryId = lh.addEntry("data".getBytes());
        }
        checkpoint = server.getBookie().journals.get(0).newCheckpoint();
        System.out.println("Checkpoint: " + checkpoint);
        assertTrue("EntryId of the recently added entry should be 0", entryId == (numOfEntries - 1));
        // We want to simulate the scenario where Bookie is killed abruptly, so
        // SortedLedgerStorage's EntryMemTable and IndexInMemoryPageManager are
        // not flushed and hence when bookie is restarted it will replay the
        // journal. Since there is no easy way to kill the Bookie abruptly, we
        // are injecting no-op shutdown.
        server.shutdown();

        conf.setDiskUsageThreshold(0.001f)
                .setDiskUsageWarnThreshold(0.0f).setReadOnlyModeEnabled(true).setIsForceGCAllowWhenNoSpace(true)
                .setMinUsableSizeForIndexFileCreation(5 * 1024);
        server = new BookieServer(conf);
        // This should have completed journal replay and force flush
        server.start();
        checkpoint = server.getBookie().journals.get(0).newCheckpoint();
        System.out.println("Checkpoint: " + checkpoint);
        server.shutdown();
        bkClient.close();
    }
```

**Comment by jvrao** [comment]
I need to look at the code to fully understand what you are saying, but what is the issue do you see here? Potential dataloss? or the log can get replayed everytime bookie restarts hence it can create duplicate entries (over-replication) ? How is this different from the following scenario? checkpoint -> receive writes -> bookie turning RO ?

**Comment by karanmehta93** [comment]
> I need to look at the code to fully understand what you are saying, but what is the issue do you see here?

The issue here is log replay multiple times. 

> Potential dataloss?

None

> or the log can get replayed everytime bookie restarts hence it can create duplicate entries (over-replication)

Yes, log will get replayed everytime. I modified the test to repeat the server restart portion of it couple times and observed that it flushes new files everytime, which can be a potential problem, given that bookie is probably out of space anyways (it has turned RO).

> How is this different from the following scenario? checkpoint -> receive writes -> bookie turning RO ?

Looking at the code implementation, only disk full, `stateManager.transitionToReadOnlyMode()` is called, which just marks the node as RO in ZK. If the bookie shuts down safely after that, it would be all good (since it flushes the logMark when shutting down), but if it crashes, it might go into this state.

**Comment by sijie** [comment]
This is fixed by #2090 and #2105
